File: examples/multi-language/scripts/python/analyze.py
import json
import os
import yaml
from datetime import datetime, timezone

# Read input from previous step
input_path = "/input/node-transform/report.json"
with open(input_path) as f:
    report = json.load(f)
print(f"Read report from {input_path}")

# Enrich with Python analysis
enriched = {
    "generatedBy": "python-analyze",
    "timestamp": datetime.now(timezone.utc).isoformat(),
    "sourceStep": "node-transform",
    "categories": {},
}

for cat, stats in report["summary"].items():
    enriched["categories"][cat] = {
        "count": stats["count"],
        "avgScore": stats["avgScore"],
        "rating": "excellent" if stats["avgScore"] >= 90 else "good" if stats["avgScore"] >= 80 else "fair",
    }

enriched["languagesByCategory"] = {}
for lang in report["ranked"]:
    cat = lang["category"]
    if cat not in enriched["languagesByCategory"]:
        enriched["languagesByCategory"][cat] = []
    enriched["languagesByCategory"][cat].append(lang["language"])

# Write both YAML and JSON outputs
yaml_path = os.path.join("/output", "enriched.yaml")
json_path = os.path.join("/output", "enriched.json")

with open(yaml_path, "w") as f:
    yaml.dump(enriched, f, default_flow_style=False, sort_keys=False)
print(f"Enriched YAML written to {yaml_path}")

with open(json_path, "w") as f:
    json.dump(enriched, f, indent=2)
print(f"Enriched JSON written to {json_path}")

print(f"Categories analyzed: {', '.join(enriched['categories'].keys())}")

File: examples/multi-language/scripts/nodejs/analyze.js
const {groupBy, mapValues, meanBy, size} = require('lodash')
const fs = require('fs')
const path = require('path')

// Sample dataset
const data = [
  {category: 'backend', language: 'Go', score: 92},
  {category: 'backend', language: 'Rust', score: 95},
  {category: 'backend', language: 'Java', score: 85},
  {category: 'frontend', language: 'TypeScript', score: 90},
  {category: 'frontend', language: 'JavaScript', score: 82},
  {category: 'data', language: 'Python', score: 94},
  {category: 'data', language: 'R', score: 78},
  {category: 'data', language: 'Julia', score: 80}
]

// Analyze using lodash
const grouped = groupBy(data, 'category')
const analysis = mapValues(grouped, items => ({
  count: size(items),
  avgScore: meanBy(items, 'score'),
  languages: items.map(i => i.language)
}))

const result = {
  generatedBy: 'node-analyze',
  timestamp: new Date().toISOString(),
  totalLanguages: data.length,
  analysis
}

const outputPath = path.join('/output', 'analysis.json')
fs.writeFileSync(outputPath, JSON.stringify(result, null, 2))
console.log(`Analysis written to ${outputPath}`)
console.log(`Categories: ${Object.keys(analysis).join(', ')}`)
